Scheduling operation for the circular buffer optimization

skeqiqevian opened the pull request.

yamaguchi1024 commented on src/exo/LoopIR_scheduling.py:
```diff
@@ -3584,6 +3584,193 @@ def mk_write(c):
+            env = {
+                s.iter: (
+                    s.lo.val if isinstance(s.lo, LoopIR.Const) else None,
+                    s.hi.val - 1 if isinstance(s.hi, LoopIR.Const) else None,
```
So this doesn't handle loops like
```
for i in seq(m, m + 3):
  ...
```
?

skeqiqevian replied:
Fixed to support this. However, currently we cannot handle general for loops with non-constant window loop bounds.

yamaguchi1024 commented on src/exo/LoopIR_scheduling.py:
```diff
@@ -3584,6 +3584,193 @@ def mk_write(c):
+        return self.bounds.pop()
+
+    def validate_access(self, idx: Optional[int]) -> bool:
+        if self.bounds is None:
```
Is this None or []?

yamaguchi1024 commented on src/exo/LoopIR_scheduling.py:
```diff
@@ -3584,6 +3584,193 @@ def mk_write(c):
+            if isinstance(w_access, LoopIR.Interval):
+                lo_rng = index_range_analysis_wrapper(w_access.lo, self.env)
+                hi_rng = index_range_analysis_wrapper(w_access.hi, self.env)
+                self.update_bounds(lo_rng | hi_rng)
```
Why is this not using `merge_index_ranges`?

skeqiqevian replied:
It doesn't have to because `index_range_analysis` will always return a valid IndexRange. `merge_index_ranges` is necessary if we don't know that the inputs might be `None` type.

yamaguchi1024 commented on src/exo/LoopIR_scheduling.py:
```diff
@@ -3584,6 +3584,193 @@ def mk_write(c):
+        return self.exit_scope()
+
+    def do_s(self, s):
+        bounds = None
```
Can you not use the same name for self.bounds and bounds?

yamaguchi1024 commented on src/exo/LoopIR_scheduling.py:
```diff
@@ -3584,6 +3586,193 @@ def mk_write(c):
+        return self.exit_scope()
+
+    def do_s(self, s):
+        print(s, self.access_window_per_scope[-1])
```
delete print?

yamaguchi1024 commented on tests/test_schedules.py:
```diff
@@ -1132,6 +1132,188 @@ def foo(a: f32 @ DRAM, b: f32 @ DRAM):
+
+#     foo = fold_buffer(foo, foo.find("x: _"), 0, 3)
+#     foo = simplify(foo)
+#     assert str(foo) == golden
```
```
x : R[N]
for i in seq(0, N):
  x[i] = 0.0
x[0] = 1.0
```

yamaguchi1024 commented on tests/test_schedules.py:
```diff
@@ -1132,6 +1132,188 @@ def foo(a: f32 @ DRAM, b: f32 @ DRAM):
+        x[3] = 0.0
+        x[2] = 0.0
+        x[7] = 0.0
+        x[5] = 0.0
```
This should succeed, right?
```
    @proc
    def foo():
        x: i8[20]
        x[0] = 0.0
        x[2] = 0.0
        x[3] = 0.0
        x[2] = 0.0
        x[7] = 0.0
        x[5] = 0.0
```
`foo = fold_buffer(foo, foo.find("x: _"), 0, 10)`

skeqiqevian commented:
@yamaguchi1024 Can you approave? 

This change will fail ExoBLAS tests because I fixed a bug with range analysis which previously enabled omitting `exo_floor_div` in some asserts. While those omissions were mathematically correct, it's not something that our current range analysis should be able to handle, so it was only able to make the omission via the bug that I fixed.

yamaguchi1024 commented:
Sounds good! @SamirDroubi might want to review, too?
